File: module-app/src/main/java/maple/expectation/global/response/ApiResponse.java
package maple.expectation.global.response;

import com.fasterxml.jackson.annotation.JsonInclude;

/**
 * API 공통 응답 포맷
 *
 * @param success 성공 여부
 * @param data 응답 데이터 (성공 시)
 * @param error 에러 정보 (실패 시)
 * @param <T> 응답 데이터 타입
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiResponse<T>(boolean success, T data, ErrorInfo error) {
  /** 성공 응답 생성 */
  public static <T> ApiResponse<T> success(T data) {
    return new ApiResponse<>(true, data, null);
  }

  /** 실패 응답 생성 */
  public static <T> ApiResponse<T> error(String code, String message) {
    return new ApiResponse<>(false, null, new ErrorInfo(code, message));
  }

  /** 에러 정보 */
  public record ErrorInfo(String code, String message) {}
}

File: module-app/src/main/kotlin/maple/expectation/response/ApiResponse.kt
package maple.expectation.response

import com.fasterxml.jackson.annotation.JsonInclude

/**
 * API common response format
 *
 * @param success success flag
 * @param response data (on success)
 * @param error error information (on failure)
 * @param T response data type
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
data class ApiResponse<T>(
    val success: Boolean,
    val data: T? = null,
    val error: ErrorInfo? = null
) {
    /** Error information record */
    data class ErrorInfo(
        val code: String,
        val message: String
    )

    companion object {
        /**
         * Create success response
         */
        fun <T> success(data: T): ApiResponse<T> = ApiResponse(true, data, null)

        /**
         * Create error response
         */
        fun <T> error(code: String, message: String): ApiResponse<T> =
            ApiResponse(false, null, ErrorInfo(code, message))

        /**
         * Create error response from ErrorCode
         */
        fun <T> error(errorCode: maple.expectation.error.ErrorCode): ApiResponse<T> =
            error(errorCode.code, errorCode.message)

        /**
         * Create error response with formatted message
         */
        fun <T> error(errorCode: maple.expectation.error.CommonErrorCode, vararg args: Any): ApiResponse<T> =
            error(errorCode.code, errorCode.formatMessage(*args))
    }
}